amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/KinesisClientFacade.java

```diff
@@ -0,0 +1,103 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import software.amazon.awssdk.services.kinesis.KinesisAsyncClient;
+import software.amazon.awssdk.services.kinesis.model.DescribeStreamSummaryRequest;
+import software.amazon.awssdk.services.kinesis.model.DescribeStreamSummaryResponse;
+import software.amazon.awssdk.services.kinesis.model.KinesisException;
+import software.amazon.awssdk.services.kinesis.model.LimitExceededException;
+import software.amazon.awssdk.services.kinesis.model.ResourceInUseException;
+import software.amazon.awssdk.services.kinesis.model.ResourceNotFoundException;
+import software.amazon.kinesis.common.KinesisRequestsBuilder;
+
+/**
+ * Facade pattern to simplify interactions with a {@link KinesisAsyncClient}.
+ */
+@Slf4j
+public final class KinesisClientFacade {
+
+    /**
+     * Reusable {@link AWSExceptionManager}.
+     * <p>
+     * N.B. This instance is mutable, but thread-safe for <b>read-only</b> use.
+     * </p>
+     */
+    private static final AWSExceptionManager AWS_EXCEPTION_MANAGER;
+
+    // FIXME dependency injection
+    private static KinesisAsyncClient kinesisClient;
+
+    static {
+        AWS_EXCEPTION_MANAGER = new AWSExceptionManager();
+        AWS_EXCEPTION_MANAGER.add(KinesisException.class, t -> t);
+        AWS_EXCEPTION_MANAGER.add(LimitExceededException.class, t -> t);
+        AWS_EXCEPTION_MANAGER.add(ResourceInUseException.class, t -> t);
+        AWS_EXCEPTION_MANAGER.add(ResourceNotFoundException.class, t -> t);
+    }
+
+    static void initialize(final KinesisAsyncClient client) {
+        kinesisClient = client;
+    }
+
+    public static DescribeStreamSummaryResponse describeStreamSummary(final String streamArn) {
+        final DescribeStreamSummaryRequest request = KinesisRequestsBuilder
+                .describeStreamSummaryRequestBuilder().streamARN(streamArn).build();
+        final CompletableFuture<DescribeStreamSummaryResponse> future = kinesisClient.describeStreamSummary(request);
+        final ServiceCallerSupplier<DescribeStreamSummaryResponse> dss =
+                () -> future.get(10, TimeUnit.SECONDS);
+        return retryWhenThrottled(dss, 3, streamArn, "DescribeStreamSummary");
+    }
+
+    // FIXME code lifted-and-shifted from FanOutConsumerRegistration; that class
+    //      (and others) should not be responsible for interacting directly with
+    //      the thread-safe Kinesis client (and handling retries, etc.)
+    private static <T> T retryWhenThrottled(
+            @NonNull final ServiceCallerSupplier<T> retriever,
+            final int maxRetries,
+            final String streamArn,
+            @NonNull final String apiName) {
+        LimitExceededException finalException = null;
+
+        int retries = maxRetries;
+        while (retries > 0) {
+            try {
+                try {
+                    return retriever.get();
```

Comment:
Dont we need to get a new future by calling kinesisClient.describeStreamSummary before retrying?

Also dont understand whats the value of nested try-catch blocks?

Reply:
Great catch w.r.t. the isolated `future = ...`! This would, indeed, reuse the singular `Future` which may have completed exceptionally.

---
> Also dont understand whats the value of nested try-catch blocks?

The inner `try` allows for Exception substitution (e.g., catch-repackage-throw, unwrap-and-throw). You'll often see an [`AWSExceptionManager`](https://github.com/awslabs/amazon-kinesis-client/blob/master/amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/AWSExceptionManager.java) in similar nested-`try`.